File: 138-copy-list-with-random-pointer/138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        
        Node h;
        HashMap<Node,Node> map=new HashMap<Node,Node>();
        h=head;
        while(h!=null)
        {
            Node newNode=new Node(h.val);
            map.put(h,newNode);
            h=h.next;
        }
        
        h=head;
        while(h!=null)
        {
            Node next=h.next;
            Node random=h.random;
            Node newNode=map.get(h);
            Node nextNewNode;
            Node randomNode;
            if(next!=null)
            {
                 nextNewNode=map.get(next); 
            }
            else
            {
                nextNewNode=null;
            }
            if(random!=null)
            {
                randomNode=map.get(random);
            }
            else
            {
                randomNode=null;
            }  
            newNode.next=nextNewNode;
            newNode.random=randomNode;
            h=h.next;
        }
        return map.get(head);
        
        
    }
}

File: 138-copy-list-with-random-pointer/138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        
        Node* newHead,*temp,*newNode,*prev,*nextRandom;
        unordered_map<Node*,Node*> mp;
        temp=head;
        while(temp!=NULL)
        {
            newNode=new Node(temp->val);
            mp.insert({temp,newNode});
            temp=temp->next;
        }
        //cout<<"here";
        newHead=NULL;
        temp=head;
        while(temp!=NULL)
        {
            newNode=mp[temp];
            if(temp->random!=NULL)
            {
                nextRandom=mp[temp->random];
            }
            else
            {
                nextRandom=NULL;
            }
            if(newHead==NULL)
            {
                newHead=newNode;
                prev=newHead;
                prev->random=nextRandom;
            }
            else
            {
                prev->next=newNode;
                prev=prev->next;
                prev->random=nextRandom;
            }
            temp=temp->next;
        }
        return newHead;
    }
};